### PR TITLE
Create a test specification package

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @Eilon \ @davidfowl \ @khellang

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+            var scopeFactory = container.GetRequiredService<IServiceScopeFactory>();
+            using (var scope = scopeFactory.CreateScope())
+            {
+                var containerScopedService = container.GetService<IFakeScopedService>();
+                var scopedService1 = scope.ServiceProvider.GetService<IFakeScopedService>();
+                var scopedService2 = scope.ServiceProvider.GetService<IFakeScopedService>();
+
+                // Assert
+                Assert.NotEqual(containerScopedService, scopedService1);
+                Assert.Equal(scopedService1, scopedService2);
```
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: `NotSame` and `Same`? Just for consistency with the other tests. I also think it conveys the intent clearer.

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+                var innerScopeFactory = outerScope.ServiceProvider.GetService<IServiceScopeFactory>();
+                using (var innerScope = innerScopeFactory.CreateScope())
+                {
+                    var outerScopedService = outerScope.ServiceProvider.GetService<IFakeScopedService>();
+                    var innerScopedService = innerScope.ServiceProvider.GetService<IFakeScopedService>();
+
+                    // Assert
+                    Assert.NotNull(outerScopedService);
+                    Assert.NotNull(innerScopedService);
+                    Assert.NotEqual(outerScopedService, innerScopedService);
```
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Same.

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+            using (var outerScope = outerScopeFactory.CreateScope())
+            {
+                var innerScopeFactory = outerScope.ServiceProvider.GetService<IServiceScopeFactory>();
+                using (var innerScope = innerScopeFactory.CreateScope())
+                {
+                    var outerScopedService = outerScope.ServiceProvider.GetService<IFakeScopedService>();
+                    var innerScopedService = innerScope.ServiceProvider.GetService<IFakeScopedService>();
+
+                    // Assert
+                    Assert.NotEqual(outerScopedService, innerScopedService);
```
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Same here.

**Comment by khellang** [comment]
LGTM :+1:

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Good catch - I copied some of the tests over without too much cleanup

**Comment by Eilon** [comment]
:shipit:

**Comment by pranavkm** [comment]
cc @tillig \ @alexmg for Autofac

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+                    // Assert
+                    Assert.NotNull(outerScopedService);
+                    Assert.NotNull(innerScopedService);
+                    Assert.NotEqual(outerScopedService, innerScopedService);
+                }
+            }
+        }
+
+        [Fact]
+        public void DisposingScopeDisposesService()
```
Comment (tillig) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Should this check for double-disposal?

**Comment by tillig** [comment]
I dig it. It'll be nice to not copy these into our test project directly.

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Would throwing from FakeService if it's already disposed suffice for this scenario? Or am I missing something?

Comment (tillig) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: I think that would work fine.

Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Done - https://github.com/aspnet/DependencyInjection/pull/312/files#diff-3f387a1ab4058855aaa78b09c60fdcc4R16

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/ActivatorUtilitiesTests.cs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection.Specification.Fakes;
+using Xunit;
+
+namespace Microsoft.Extensions.DependencyInjection.Specification
+{
+    public abstract partial class DependencyInjectionSpecificationTests
```
Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/ActivatorUtilitiesTests.cs]: @Eilon how much do you hate this? It seemed kinda nice to be able to have one test that could be sublcassed to get the full test coverage. However, dumping the `ActivatorUtilitiesTests` in the other class made it pretty hard to navigate it.

Comment (Eilon) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/ActivatorUtilitiesTests.cs]: :disappointed: but whatever.

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+    {
+        protected abstract IServiceProvider CreateServiceProvider(IServiceCollection serviceCollection);
+
+        [Fact]
+        public void ServicesRegisteredWithImplementationTypeCanBeResolved()
+        {
+            // Arrange
+            var collection = new ServiceCollection();
+            collection.AddTransient(typeof(IFakeService), typeof(FakeService));
+            var container = CreateServiceProvider(collection);
```
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Nit: Why did these stay `container` when the method became `CreateServiceProvider`?

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+            var service2 = container.GetService<IFakeService>();
+
+            // Assert
+            Assert.IsType<FakeService>(service1);
+            Assert.IsType<FakeService>(service2);
+            Assert.NotSame(service1, service2);
+        }
+
+        [Fact]
+        public void ServicesRegisteredWithImplementationType_ReturnSaneInstancesPerResolution_ForSingletons()
```
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: `Sane` or `Same`?

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+            // Act
+            var service = container.GetService<IFakeService>();
+
+            // Assert
+            Assert.NotNull(service);
+            Assert.IsType<FakeService>(service);
+        }
+
+        [Fact]
+        public void ServicesRegisteredWithImplementationType_ReturnDifferentInstancesPerResolution_ForTransientServices()
```
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: I think this test should be expanded to verify that transients are unique when resolved from a different scope as well. Not just the root provider. Either that or add a different test.

Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: We could add a variant to `ScopedServiceCanBeResolved` to include transients.

Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Ah, included a scoped resolve in #313. Want me to move it?

Comment (pranavkm) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Yeah - would be nice to cover it as two separate scenarios. Would make it easier to diagnose what the issue is if it happens to break.

Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Alright. Updated.

**Pull request closed without merging**